Improve RP2 DVP PIO driver to work with faster cameras, like OV5640

sfe-SparkFro opened the pull request:
Fixes #1

Some cameras, like the OV5640, output rows of pixels in very short bursts, with long gaps between each row. Below is an example, where the green trace is the HSYNC signal (high during each row burst):

<img width="1280" height="800" alt="image" src="https://github.com/user-attachments/assets/aae6a7f6-b198-48f9-9e81-d8b3a5c4925d" />

Although the average data rate is easily handled by the QSPI bus, each row burst can be faster than the max QSPI bus rate. I had previously thought the XIP cache would buffer the data by itself, but apparently not, resulting in missed data:

<img width="320" height="240" alt="image" src="https://github.com/user-attachments/assets/ef1bcbb8-afad-4479-b351-a56426bc4791" />

This changes the driver to buffer each row of pixels in SRAM, then have another DMA slowly send the row to PSRAM using a control block architecture, similar to the [HSTX DVI driver](https://github.com/sparkfun/red_vision/blob/ea793c7582f2245f9788e86ca5f767b68e9128d0/red_vision/displays/dvi_rp2_hstx.py#L430):

<img width="505" height="290" alt="image" src="https://github.com/user-attachments/assets/54525345-f38d-4136-8976-4b7f6349156c" />

Now the driver can capture clean images from the OV5640 at 30 FPS:

<img width="320" height="240" alt="image" src="https://github.com/user-attachments/assets/34c2cbc8-5254-43d1-a5b7-8f94c9126136" />

Also seems to fix #2, possibly because this no longer uses interrupts.

sfe-SparkFro committed:
Improve RP2 DVP PIO driver

Fixes #1
Now buffers each row of pixels in SRAM if needed. DMA setup is similar to HSTX DVI driver

sfe-SparkFro committed:
Change default OV5640 XCLK frequency

20MHz XCLK results in 30MHz PCLK, which the PIO can just barely handle (37.5MHz max).
25MHz XCLK results in 40MHz PCLK, which is too fast.
Could also change PLL settings, but this is simpler.

gigapod reviewed (approved):
Looks good to me. I don't see anything that needs  changing.

The only possible nit would be changing some of the hard coded numbers (constants, memory sizes ...etc) to actual named constants not "magic numbers", but really nothing to worry about.

malcolm-sparkfun reviewed (approved):
LGTM

sfe-SparkFro committed:
Add optional `continuous` paramter to RP2 DVP cameras

There are some edge-case issues with having the DMAs automatically
capture every frame from the camera. This change makes it default to
non-continuous mode, but users can enable it if they really want.
